### ISSUE TITLE
Server sometimes receives and misinterprets request response

### ISSUE THREAD

**Issue by rosalogia** [marigold-dev/pollinate#2]
Currently the server and the client share a socket, and the server calls `recvfrom` asynchronously while the client is at liberty to make requests and receive responses to requests via a send and receive API. There are two big issues with this

1. When the client makes a request a peer, there's nothing stopping the server's call to `recvfrom` from receiving the response from the peer.
2. When serializing a value of a sum type (like our request and response types in our tests), Bin_prot does **not** preserve type information. Thus, in our tests, when the server unpacks the serialized payload, it has no way of knowing whether it's unpacking a response or a request.

Some ways we could resolve this include

1. Having a separate socket for the client and the server, I like this solution the least. I tried implementing it and it results in an annoying amount of code duplication.
2. Simply ensuring that requests and responses are not represented the way they currently are in our tests. One alternative that would work is simply wrapping requests and responses in a message sum type:
   ```ocaml
    type message =
      | Request of Request.t
      | Response of Response.t
    ```
   This way, whether or not the message is a request or a response would be encoded in the serialized payload. I don't feel too strongly about this solution either, since it places a restriction on the end user.
3. Creating a message type as specified above _in_ the library, but with requests and responses merely being `bytes`. This way we can ensure that for our purposes, whether the message is a request or a response is encoded in the serialized payload, but abstract this away from the consumer. This solution might be clean, but I'll have to give it a try and see.
4. If we really must, we can dive into the depths of the `Lwt_unix` module and see how we can set up polling and blocking, that might solve our problems but I honestly don't know yet :smile:

**Comment by gsebil08** [comment]
>Currently the server and the client share a socket

I may be wrong, but since each peer must be able to send **and** receive messages, by design they will have one server and one client sharing the same socket, am I right?

>When serializing a value of a sum type (like our request and response types in our tests), Bin_prot does not preserve type information.

What about opening an issue on https://github.com/janestreet/bin_prot to ask them? 🤔 
Moreover, I think it would be useful for such a feature to be within `bin_prot` standard library rather that in every project of people encoding sum type, don't you think?

**Comment by rosalogia** [comment]
> I may be wrong, but since each peer must be able to send and receive messages, by design they will have one server and one client sharing the same socket, am I right?

Yeah, I mean it's possible to implement with two sockets but it's not the norm and it doesn't make a lot of sense. I'd rather not do that.

> What about opening an issue on https://github.com/janestreet/bin_prot to ask them? thinking
Moreover, I think it would be useful for such a feature to be within bin_prot standard library rather that in every project of people encoding sum type, don't you think?

We can open an issue, but it would have to be a feature request. This isn't necessarily a bug in Bin_prot but rather a design choice.

I wonder actually if maybe our library should just be agnostic to the binary encoding protocol as well. In the end, we are always giving raw bytes to the consumer and simply giving them some utilities to pack/unpack their data with Bin_prot, so I wonder is it necessary to make Bin_prot a part of this library? :thinking:
